rosbag2_py/test/test_transport.py

```diff
@@ -80,6 +81,170 @@ def test_recoder_log_level(tmp_path, storage_id):
+    assert player.play_next()
+    assert 2 == player.burst(2)
+    player.seek(0)
+    # This requires that the player node be spun by an executor
+    resume_client.call(Resume.Request(), 5.0)
+    assert not player.play_next()
+    assert player.wait_for_playback_to_finish(15.0)
+    player.stop()
+
+    # 2 msgs with play_next, 2 msgs with burst, then seek to the beginning and play all 10 msgs
+    expected_number_of_messages = 14
+    assert wait_for(
+        lambda: len(msgs) == expected_number_of_messages,
+        timeout=rclpy.duration.Duration(seconds=10),
+    ), str(msgs)
+
+    executor.shutdown()
+    executor_thread.join(3)
+    assert not executor_thread.is_alive()
+    resume_client.destroy()
+    sub.destroy()
+    node.destroy_node()
+    ctx.shutdown()
+
+
+@pytest.mark.parametrize('storage_id', TESTED_STORAGE_IDS)
+def test_recorder_api(tmp_path, storage_id):
+    bag_path = tmp_path / 'test_recorder_api'
+    storage_options, _ = get_rosbag_options(str(bag_path), storage_id)
+
+    record_options = rosbag2_py.RecordOptions()
+    record_options.start_paused = False
+    record_options.topics = ['/topic']
+    record_options.is_discovery_disabled = False
+    record_options.topic_polling_interval = datetime.timedelta(milliseconds=10)
+    record_options.disable_keyboard_controls = True
+
+    recorder = rosbag2_py.Recorder(
+        storage_options, record_options, 'info', 'rosbag2_recorder_test')
+    recorder.start_spin()
+    recorder.record()
+
+    ctx = rclpy.Context()
+    ctx.init()
+    node = rclpy.create_node('test_recorder_api_node', context=ctx)
+    pub = node.create_publisher(String, 'topic', QoSProfile(depth=10))
+    resume_client = node.create_client(Resume, '/rosbag2_recorder_test/resume')
+    executor = rclpy.executors.SingleThreadedExecutor(context=ctx)
+    executor.add_node(node)
+    executor_thread = threading.Thread(
+        target=executor.spin,
+        daemon=True)
+    executor_thread.start()
+
+    assert resume_client.wait_for_service(5.0), str(node.get_service_names_and_types())
+    recorder.resume()
+    recorder.pause()
+    # This requires that the recorder node be spun by an executor
+    resume_client.call(Resume.Request(), 5.0)
+    i = 0
+    while rclpy.ok() and i < 10:
+        pub.publish(String(data=str(i)))
+        i += 1
+    recorder.stop()
+
+    executor.shutdown()
+    executor_thread.join(3)
+    assert not executor_thread.is_alive()
+    resume_client.destroy()
+    pub.destroy()
+    node.destroy_node()
+    ctx.shutdown()
+
+    metadata_io = rosbag2_py.MetadataIo()
+    assert wait_for(
+        lambda: metadata_io.metadata_file_exists(str(bag_path)),
+        timeout=rclpy.duration.Duration(seconds=10))
+    metadata = metadata_io.read_metadata(str(bag_path))
+    bag_topics = [
+        topic_info.topic_metadata.name
+        for topic_info in metadata.topics_with_message_count
+    ]
+    assert '/topic' in bag_topics, str(bag_topics)
```

Comment:
I believe the metadata file should exist & contain the `/topic` topic after `recorder.stop()`: actually publishing on the topic isn't really required here because the topic will exist in the metadata, since the topic was discovered by the recorder and a writer was created.

I am not checking the _number_ of messages here to avoid races or awkwardly recording & stopping multiple times until the metadata contains the correct number of messages.